### PR TITLE
Fix `valueFrom` bug in EnvVars

### LINKED ISSUE #278: Env var is passed to Runner Pod, but NOT into terraform binary
Hi Guys,
I've hit an issue I can't resolve for few hours.
According to #232 and #223 env vars are injected into runner pod.
So I added following line to Terraform custom resource:
```
  runnerPodTemplate:
    spec:
      env:
        - name: VAULT_TOKEN
          valueFrom:
            secretKeyRef:
              name: vault-token
              key: token
```

Whenever runner start it's duty and call `terraform plan` I see in logs such lines and plan fails:
```
0804 14:25:28.653503       8 request.go:601] Waited for 1.046720264s due to client-side throttling, not priority and fairness, request: GET:https://172.16.64.1:443/apis/storage.k8s.io/v1beta1?timeout=32s
{"level":"info","ts":"2022-08-04T14:25:29.382Z","logger":"runner.terraform","msg":"preparing for Upload and Extraction"}
{"level":"info","ts":"2022-08-04T14:25:29.383Z","logger":"runner.terraform","msg":"write backend config","path":"/tmp/terraform-operator-terraform-kv-engine-weaver/live/devopsprod000/engines/key-value","config":"generated_backend_config.tf"}
{"level":"info","ts":"2022-08-04T14:25:29.383Z","logger":"runner.terraform","msg":"write config to file","filePath":"/tmp/terraform-operator-terraform-kv-engine-weaver/live/devopsprod000/engines/key-value/generated_backend_config.tf"}
{"level":"info","ts":"2022-08-04T14:25:29.384Z","logger":"runner.terraform.runner.terraform","msg":"looking for path","file":"terraform"}
{"level":"info","ts":"2022-08-04T14:25:29.384Z","logger":"runner.terraform","msg":"creating new terraform","workingDir":"/tmp/terraform-operator-terraform-kv-engine-weaver/live/devopsprod000/engines/key-value","execPath":"/usr/local/bin/terraform"}
{"level":"info","ts":"2022-08-04T14:25:29.385Z","logger":"runner.terraform","msg":"setting envvars"}
{"level":"info","ts":"2022-08-04T14:25:29.385Z","logger":"runner.terraform","msg":"getting envvars from os environments"}
{"level":"info","ts":"2022-08-04T14:25:29.386Z","logger":"runner.terraform","msg":"initializing"}
{
  "terraform_version": "1.1.9",
  "platform": "linux_amd64",
  "provider_selections": {},
  "terraform_outdated": false
}

Initializing the backend...

Successfully configured the backend "kubernetes"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding latest version of hashicorp/vault...
- Installing hashicorp/vault v3.8.0...
- Installed hashicorp/vault v3.8.0 (signed by HashiCorp)

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!
{"level":"info","ts":"2022-08-04T14:25:30.242Z","logger":"runner.terraform","msg":"setting up the input variables"}
{"level":"info","ts":"2022-08-04T14:25:30.243Z","logger":"runner.terraform","msg":"mapping the Spec.Vars"}
{"level":"info","ts":"2022-08-04T14:25:30.243Z","logger":"runner.terraform","msg":"mapping the Spec.VarsFrom"}
{"level":"info","ts":"2022-08-04T14:25:30.257Z","logger":"runner.terraform","msg":"creating a plan"}

Error: no vault token found

  with provider["registry.terraform.io/hashicorp/vault"],
  on main.tf line 1, in provider "vault":
   1: provider "vault" {

{"level":"error","ts":"2022-08-04T14:25:30.799Z","logger":"runner.terraform","msg":"error creating the plan","error":"exit status 1\n\nError: no vault token found\n\n  with provider[\"registry.terraform.io/hashicorp/vault\"],\n  on main.tf line 1, in provider \"vault\":\n   1: provider \"vault\" {\n\n"}
{"level":"info","ts":"2022-08-04T14:25:30.800Z","logger":"runner.terraform","msg":"cleanup TmpDir","tmpDir":"/tmp/terraform-operator-terraform-kv-engine-weaver"}

```

If I exec into the runner pod and call:
```
echo $VAULT_TOKEN
```
I can see correct value.
Also calling manually `terraform init, plan and apply` from the pod context works good!

Therefore, I'm in the opinion that env var is indeed injected into the pod, but NOT into the terraform process/binary that is triggered as no 1 process (tini).

I'm using:
- helm chart tf-controller version 0.3.8
- terraform 1.2.2
- flux2 version 0.31.5
- kubernetes AKS 1.22.6

Happy to hear your thoughts and tips Guys!

thanks
Tomek

### PR DESCRIPTION
**Pull request by Nalum**
The code was setup to write any EnvVar provided in the Terraform Spec to the Runner `SetEnv` function, so if you set `valueFrom` it would still look for `value` and set that which would result in an empty string being set.

Tests added to verify that `valueFrom` is working with a `Secret` and a `ConfigMap` `valueFrom` setting.

Fixes #278

### CODE REVIEW

File: controllers/terraform_controller.go [hunk 1]
```diff
@@ -751,7 +752,25 @@ terraform {
 	envs := map[string]string{}
 
 	for _, env := range terraform.Spec.RunnerPodTemplate.Spec.Env {
-		envs[env.Name] = env.Value
+		if env.ValueFrom != nil {
+			if env.ValueFrom.SecretKeyRef != nil {
+				secret := corev1.Secret{}
+				r.Client.Get(ctx, types.NamespacedName{
```
Comment (chanwit) [inline controllers/terraform_controller.go:758]: Please handle error here

File: controllers/terraform_controller.go [hunk 2]
```diff
@@ -751,7 +752,25 @@ terraform {
...
+					Namespace: terraform.GetObjectMeta().GetNamespace(),
+					Name:      env.ValueFrom.SecretKeyRef.Name,
+				}, &secret)
+				envs[env.Name] = string(secret.Data[env.ValueFrom.SecretKeyRef.Key])
+			} else if env.ValueFrom.ConfigMapKeyRef != nil {
+				cm := corev1.ConfigMap{}
+				r.Client.Get(ctx, types.NamespacedName{
```
Comment (chanwit) [inline controllers/terraform_controller.go:765]: Please also handle error here too

**Comment by Nalum** [comment]
Thanks @chanwit I've made changes to handle the `error`s you pointed out.

**Review by chanwit — APPROVED** [review]

**Comment by chanwit** [comment]
LGTM thank you @Nalum

**Pull request merged**